### DISCUSSION TITLE
Support Hot Reload

### DISCUSSION THREAD

**Discussion by zvictor** [pulumi/pulumi#11065] [Ideas]
I have been using `aws.lambda.Function` to deploy to my _localstack_ setup and I was hoping there would be a fast way to iterate over local changes, but I couldn't find a way to do so.

I noticed that `aws.lambda.Function` zips the code (`pulumi.asset.AssetArchive`), which means that enabling _hot swapping_ in _localstack_ won't make any difference unless we have a way to make Pulumi repack the content of the lambda on-demand.

* I have tried using [pulumi watch](https://www.pulumi.com/docs/reference/cli/pulumi_watch/), but with no success. It seems that only the pulumi code is being watched, not the code of the lambda functions.

* I have also tried running pulumi up every time a change was made on the lambda code, but that has extreme bad performance and possibly other side effects.

Thus, I would like to suggest that create ways in pulumi to trigger a hot reload of the code of affected deployments.

**Comment by zvictor** [comment]
I have also posted this issue as a question on both [Slack](https://pulumi-community.slack.com/archives/C84L4E3N1/p1666088263479549) and [Stackoverflow](https://stackoverflow.com/q/74121832/599991) to make sure I am not missing something here. No reactions on any of them so far.

**Reply by Frassle** [reply]
I don't believe we have anything to support this today.

**Comment by iulspop** [comment]
What if you tried [Pulumi's "magic" Lambda function feature](https://www.pulumi.com/docs/guides/crosswalk/aws/lambda/#structuring-your-serverless-code-base) in combination with [`pulumi watch`](https://www.pulumi.com/docs/reference/cli/pulumi_watch/). Write your Lambda function handlers as normal (have to type the handler function using the Pulumi `aws.lambda.EventHandler` type as in [this example](https://github.com/pulumi/examples/blob/master/aws-ts-apigateway-lambda-serverless/handler.ts)), then import it like you would a normal module and use it to define a Lambda function:
```
import handler from "./handler";

new aws.lambda.CallbackFunction("mylambda", {
  callback: hanlder;
})
```

Since the Lambda function handler is now part of the Pulumi code, `pulumi watch` should be triggered and update the resources. 

However, the [function serialization feature](https://www.pulumi.com/docs/intro/concepts/function-serialization/#function-serialization) is only supported in TS & JS.

If you use a different language or don't want to use the serialization feature, perhaps changing the directory watched by `pulumi watch` using the [`--cwd` flag](https://www.pulumi.com/docs/reference/cli/pulumi_watch/#:~:text=%2DC%20or-,%2D%2Dcwd%20flag,-to%20use%20a) to include the Lambda function files would trigger updating those ressources?
